consumer/src/main/kotlin/streams/StreamsEventSinkExtensionFactory.kt

```diff
@@ -37,54 +37,52 @@ class StreamsEventSinkExtensionFactory : KernelExtensionFactory<StreamsEventSink
-                        val streamsQueryExecution = StreamsEventSinkQueryExecution(streamsTopicService, db,
-                                logService.getUserLog(StreamsEventSinkQueryExecution::class.java),
-                                strategyMap)
+                        try {
+                            streamsLog.info("Initialising the Streams Sink module")
+                            val streamsSinkConfiguration = StreamsSinkConfiguration.from(configuration)
+                            val streamsTopicService = StreamsTopicService(db)
+                            val strategyMap = TopicUtils.toStrategyMap(streamsSinkConfiguration.topics,
+                                    streamsSinkConfiguration.sourceIdStrategyConfig)
+                            val streamsQueryExecution = StreamsEventSinkQueryExecution(streamsTopicService, db,
+                                    logService.getUserLog(StreamsEventSinkQueryExecution::class.java),
+                                    strategyMap)
 
-                        // Create the Sink
-                        val log = logService.getUserLog(StreamsEventSinkFactory::class.java)
-                        eventSink = StreamsEventSinkFactory
-                                .getStreamsEventSink(configuration,
-                                        streamsQueryExecution,
-                                        streamsTopicService,
-                                        log,
-                                        db)
-                        // start the Sink
-                        if (Neo4jUtils.isCluster(db)) {
-                            log.info("The Sink module is running in a cluster, checking for the ${Neo4jUtils.LEADER}")
-                            Neo4jUtils.waitForTheLeader(db, log) { initSinkModule(streamsTopicService, streamsSinkConfiguration) }
-                        } else {
-                            // check if is writeable instance
-                            Neo4jUtils.executeInWriteableInstance(db) { initSinkModule(streamsTopicService, streamsSinkConfiguration) }
-                        }
+                            // Create the Sink
+                            val log = logService.getUserLog(StreamsEventSinkFactory::class.java)
+                            eventSink = StreamsEventSinkFactory
+                                    .getStreamsEventSink(configuration,
+                                            streamsQueryExecution,
+                                            streamsTopicService,
+                                            log,
+                                            db)
+                            // start the Sink
+                            if (Neo4jUtils.isCluster(db)) {
+                                log.info("The Sink module is running in a cluster, checking for the ${Neo4jUtils.LEADER}")
+                                Neo4jUtils.waitForTheLeader(db, log) { initSinkModule(streamsTopicService, streamsSinkConfiguration) }
```

Comment:
this is at startup, how do we do here with leader switches? if the sink module is not initialized on the non-leaders?

Reply:
this only waits for the leader election and then executes the `initSinkModule` on all cluster members.
Maybe we can do this better:
https://github.com/neo4j-contrib/neo4j-streams/blob/599a0813a456c89d741b6822f36a2d31916de2a5/consumer/src/main/kotlin/streams/StreamsEventSinkExtensionFactory.kt#L82-L85

by executing line 83/84 in only in a writeable instance?

Reply:
would it wait forever? so possibly blocking the db startup for the non-leader members?
Or what happens on the non-leaders?

I guess starting it on the writeable instances makes sense.

Reply:
No, it only checks if a leader is elected, as you underscored we need to start the plugin in all instance otherwise in case of rebalancing, it could not work; this means that the `eventSink.start()` must happen in all instances. We need to do that because of the lack of a trigger in changing of cluster role.

Reply:
+1